File: 03/main.rb
passwords = File.readlines(__dir__ + '/input.txt').map do |l|
  parts = l.split(' ')
  scheme = parts[0].split('-')
  min = scheme[0].to_i
  max = scheme[1].to_i
  letter = parts[1].split(':')[0]
  password = parts[2]
  {min: min, max: max, letter: letter, password: password}
end

sum = 0

passwords.each do |pass|
  appearances = pass[:password].count(pass[:letter])
  if appearances >= pass[:min] && appearances <= pass[:max]
    sum += 1
  end
end

puts sum

File: 03/main.js
const fs = require('fs');

// Reads the input file and splits it into lines
function read(filename) {
  return fs.readFileSync(filename).toString().split(/\n/);
}

// Converts the lines of the input into objects representing the
// min, max, letter, and password contents of each line.
function clean(data) {
  return data.map(e => {
    const parts = e.split(' ');

    const [min, max] = parts[0].split('-').map(e => parseInt(e));

    const letter = parts[1].replace(':', '');

    const password = parts[2];

    return {min, max, letter, password};
  })
}

// Iterates through the array of objects, counts the number of *letter* values
// in *password* is between *min* and *max*, inclusive and if so increments
// a *sum* return value.
function count(data) {
  let sum = 0;

  data.forEach(d => {
    const letterCount = (d.password.match(new RegExp(d.letter, 'g')) || []).length;
    if (letterCount >= d.min && letterCount <= d.max) {
      sum += 1
    }
  })

  return sum;
}

(function main() {
  const data = clean(read(__dirname + '/input.txt'));
  console.log(count(data))
})();
